Min # Jsonparser step1

undervineg opened the pull request:
검토 부탁드립니다~ :)

undervine committed:
Draft struct JSONData

undervine committed:
Draft struct JSONData

undervine committed:
Implement init() of JSONData

undervine committed:
Add split() func to String extension

undervine committed:
Implement parse() func in JSONData

undervine committed:
Implement JSONData

Make regular expression of number, string, bool.
Add parse() func that convert String to number or bool or string.
Add init() that initiate each data in JSONData.
Extend String to add split() func that finds certain pattern within a String.
Add convert() to String extension that typecast each substring that come out of matched string.

undervine committed:
Add askFor() func and isJSONPattern() func

Add askFor() func that prints user menu and get user input.
Add isJSONPattern() func that check if the user input conforms JSON pattern.

undervine committed:
Add printDataReport() func

undervine committed:
Implement main() func

godrm commented on JSONParser/JSONParser/main.swift:
```diff
@@ -2,9 +2,22 @@
+    do {
+        while let userInput = try InputView.askFor(message: "분석할 JSON 데이터를 입력하세요: ") {
+            let parsedData = try JSONData.parse(from: userInput)
+            let jsonData = JSONData(parsedData)
```
이 정도면 parse 해서 객체 인스턴스를 만드는 Factory Pattern 형태로 만드는게 좋을 것 같습니다.
팩토리 패턴을 만들어서 개선하세요.

undervine committed:
Add JSONParser and move some funcs from JSONData

Modify parse() to return JSONData.
Modify split() func in String extension to 'matches()'.
Move some of matches() role into JSONParser struct because it's not the role of String to convert substrings to each type of JSONData.

undervineg commented:
음.. 제가 팩토리 패턴을 잘 이해를 못한 것 같은데 .. 다음과 같이 수정해보았습니다.
- JSONParser 구조체 추가. (기존의 JSONData 함수들 옮김)
- JSONParser의 parse() 반환타입 변경 : [Any] ➔ JSONData 
- String extension 내 기존 split() 함수 역할 축소 : String에서 특정 패턴을 찾아 JSONData에 해당하는 타입으로 변환 ➔ String에서 특정 패턴의 NSTextCheckingResult만 반환.
- JSONParser에서 NSTextCheckingResult를 JSONData에 해당하는 타입으로 변환.

undervine committed:
Modify struct JSONParser not to use RegEXP

It's inefficient to use RegEXP not only the length of code, but also the increasing number of RegExp for each type.
Therefore, move String extension to InputView only to use RegEXP for json pattern checking of user input.
Instead, add searchJsonData() to get elements of String in json standard.
Modify parse() func to use optional binding when casting type.

undervine committed:
Add invalidPattern error handling

undervine committed:
Add printError() func

undervine committed:
Add error handling

undervine committed:
Modify unit test

undervineg commented:
- 정규표현식을 쓰는 게 오히려 비효율적인 것 같아서 문자열 자르기로 변경하였습니다.
- JSON 규격 관련 에러 처리 로직을 추가하였습니다.

godrm commented on JSONParser/JSONParserTests/JSONParserTests.swift:
```diff
@@ -11,7 +11,15 @@ import XCTest
+    var parsedData: JSONData!
+    
+    override func setUp() {
+        parsedData = JSONParser.parse(from: jsonString)
```
테스트 함수들은 객체의 메소드 하나에 적어도 테스트 하나를 작성하는게 보통입니다.
setUp()에서는 전체 테스트 함수에서 필요한 객체를 만드는거고
이렇게 JSONParser.parse()를 테스트한다면 아래 테스트 함수에서 호출하고 테스트하는게 좋습니다.

undervineg replied:
네 알겠습니다~! ㅎㅎ

undervine committed:
Move global variable into each test functions

godrm commented:
수고하셨습니다. 다음 단계에서도 코드를 개선하면서 함께 테스트 코드도 개선해보세요.
